File: scala-helloworld/src/main/scala/br/edu/ufcg/ic/akka/java/FutureActor.java
package br.edu.ufcg.ic.akka.java;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.TimeUnit;
import static akka.pattern.Patterns.ask;
import static akka.pattern.Patterns.pipe;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Inbox;
import akka.actor.Props;
import akka.actor.Terminated;
import akka.actor.UntypedActor;
import akka.dispatch.Futures;
import akka.dispatch.Mapper;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Creator;
import akka.util.Timeout;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;

public class FutureActor extends UntypedActor {
	
	static public class Result {
        private final String x;
        private final String s;
        
        public Result(String x, String s) {
            this.x = x;
            this.s = s;
        }
    }
	
	public static Props props() {
        return Props.create(new Creator<FutureActor>() {
            private static final long serialVersionUID = 1L;

            @Override
            public FutureActor create() throws Exception {
                return new FutureActor();
            }

        });
    }
	
    LoggingAdapter log = Logging.getLogger(getContext().system(), this);

	@Override
	public void onReceive(Object message) throws Throwable {
		if (message instanceof String) {
            log.info("Received String message: {}", message);
            getSender().tell(message, getSelf());
        } else{
        	unhandled(message);
        }
	}
	
	public static void main(String args[]) {
	
		final ActorSystem system = ActorSystem.create("MySystem");
		final ActorRef actorA = system.actorOf(Props.create(MyUntypedActor.class),"actorA");
		final ActorRef actorB = system.actorOf(Props.create(MyUntypedActor.class),"actorB");
		final ActorRef actorC = system.actorOf(Props.create(MyUntypedActor.class),"actorC");
		
		final Timeout t = new Timeout(Duration.create(5, TimeUnit.SECONDS));

		final ArrayList<Future<Object>> futures = new ArrayList<Future<Object>>();
		futures.add(ask(actorA, "request", 1000)); // using 1000ms timeout
		futures.add(ask(actorB, "another request", 1000)); // using timeout from
		// above
	
		final Future<Iterable<Object>> aggregate = Futures.sequence(futures,
		    system.dispatcher());
	
		final Future<Result> transformed = aggregate.map(
		    new Mapper<Iterable<Object>, Result>() {
		        public Result apply(Iterable<Object> coll) {
		            final Iterator<Object> it = coll.iterator();
		            final String x = (String) it.next();
		            final String s = (String) it.next();
		            return new Result(x,s);
		        }
		    }, system.dispatcher());//novo futuro
		pipe(transformed, system.dispatcher()).to(actorC);
		
		/*system.stop(actorA);
    	system.stop(actorB);
    	system.stop(actorC);
    	system.shutdown();*/
	}
}

File: scala-helloworld/src/main/scala/br/edu/ufcg/ic/akka/FutureActor.scala
package br.edu.ufcg.ic.akka

import akka.actor.Actor
import akka.actor.ActorSystem
import akka.actor.Props
import akka.pattern.{ ask, pipe }
import scala.concurrent.Future;
import akka.util.Timeout;
import scala.concurrent.duration.FiniteDuration

object FutureActor extends App {

  future()

  final case class Result(x: Int, s: String, d: Double)
  case object Request
  
  class FutureActor extends Actor {  
        
    def receive = {
      case Request =>
        println("\n\t receive requeste from: \t\t%s\n\t".format(sender()))
        try {
          val result = Request
          sender() ! result
        } catch {
          case e: Exception =>
            sender() ! akka.actor.Status.Failure(e)
            throw e
        }

      case Result(x, s, d) =>
        println("\n\t receive result from: \t\t%s\n\t".format(sender()))
    }
  }

  def future(){
    val system = ActorSystem("system")
    import system.dispatcher // The ExecutionContext that will be used
    implicit val timeout = Timeout(FiniteDuration(5, "seconds")) // needed for `?` below
    
    val actorA = system.actorOf(Props[FutureActor], "actora")
    val actorB = system.actorOf(Props[FutureActor], "actorb")
    val actorC = system.actorOf(Props[FutureActor], "actorc")
    val actorD = system.actorOf(Props[FutureActor], "actord")

    val f: Future[Result] =
        for {
            x <- ask(actorA, Request).mapTo[Int] // call pattern directly
            s <- (actorB ask Request).mapTo[String] // call by implicit conversion
            d <- (actorC ? Request).mapTo[Double] // call by symbolic name
        } yield Result(x, s, d)
    
    //f pipeTo actorD .. or .. pipe(f) to actorD
    pipe(f) to actorD
    println(f.foreach(println))
  }
}

